### PR TITLE
Cleanup refactoring around DefPath handling

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR makes two big changes:
* All DefPaths of a crate are now stored in metadata in their own table (as opposed to `DefKey`s as part of metadata `Entry`s. 
* The compiler will no longer allocate a pseudo-local DefId for inlined HIR nodes (because those are gross). Inlined HIR nodes will have a NodeId but they don't have there own DefId anymore. Turns out they were not needed anymore either. Hopefully HIR inlining will be gone completely one day but if until then we start needing to be able to map inlined NodeIds to original DefIds, we can add an additional table to metadata that allows for reconstructing this.

Overall this makes for some nice simplifications and removal of special cases.

r? @eddyb 

cc @rust-lang/compiler

### CODE REVIEW

File: src/librustc/hir/map/def_collector.rs [hunk 1]
```diff
@@ -47,25 +47,6 @@ impl<'a> DefCollector<'a> {
...
-        let root_path = Box::new(InlinedRootPath {
-            data: parent_def_path.data,
-            def_id: parent_def_id,
-        });
-
-        let def = collector.create_def(parent_node, DefPathData::InlinedRoot(root_path));
-        collector.parent_def = Some(def);
-
-        collector
-    }
```
Comment (eddyb) [inline src/librustc/hir/map/def_collector.rs]: You forgot to remove the `hir::intravisit::Visitor` impl.

File: src/librustc/hir/map/mod.rs [hunk 2]
```diff
@@ -936,7 +940,7 @@ pub fn map_crate<'ast>(forest: &'ast mut Forest,
         forest: forest,
         dep_graph: forest.dep_graph.clone(),
         map: RefCell::new(map),
-        definitions: RefCell::new(definitions),
+        definitions: definitions,
         local_node_id_watermark: local_node_id_watermark,
         local_def_id_watermark: local_def_id_watermark,
```
Comment (eddyb) [inline src/librustc/hir/map/mod.rs]: You can remove this watermark too btw.

**Review by eddyb — APPROVED** [review]
r=me with comments addressed.

**Comment by michaelwoerister** [comment]
@bors r=eddyb

Thanks for the review!

**Comment by frewsxcv** [comment]
Travis found an ICE.

**Comment by michaelwoerister** [comment]
Something seems to break in rustdoc...

**Comment by michaelwoerister** [comment]
@bors r=eddyb

Re-reviewed.

**Pull request merged**